### PR TITLE
Test an administrator can upgrade partitioned heap table.

### PR DESCRIPTION
**Pull request by berlin-ab**
Adds test for subpartitioned heap table.

- extracts library for making assertions about rows defined as structs. Tests using the library need to define a `matcher()` function and a `match_failed()` function.

### CODE REVIEW

**Comment by dkrieger-pvtl** [comment]
administrative note: you can also open this as a [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+anAdministratorPerformsAnUpgrade(void)
+{
+	performUpgrade();
+}
+
+static void
+theSubpartitionShouldExistWithDataInTheGpdbSixCluster(void)
+{
+	Rows *rows_in_partition_a = queryForRows("select id, age from users_1_prt_partition_id_2_prt_subpartition_age_first");
+	Rows *rows_in_partition_b = queryForRows("select id, age from users_1_prt_partition_id_2_prt_subpartition_age_second");
```
Comment (dgkimura) [inline contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Any reason we can't just tack on a where clause (e.g. id=1 and age=10, etc) and then assert we get back 1 row, 2 rows, or no rows?  Seems like it could be much simpler than adding test code to parse and check the rows from PGresult.

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	int number_of_rows = PQntuples(result);
+
+	Rows *rows = calloc(1, sizeof(Rows));
+
+	int id_column_index  = PQfnumber(result, "id");
+	int age_column_index = PQfnumber(result, "age");
+
+	for (int i = 0; i < number_of_rows; i++)
+	{
+		User *user = calloc(1, sizeof(User));
```
Comment (dgkimura) [inline contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Do we care about freeing this later?

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+typedef struct UserData
+{
+	int id;
+	int age;
+} User;
+
+typedef struct RowsData
+{
+	int  size;
+	User *users[10];
```
Comment (dgkimura) [inline contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Should we assert somewhere that `number_of_rows` is not more than 10?  Else if this test is updated later without meeting that assumption, it may be a lot harder to debug.

File: contrib/pg_upgrade/test/integration/scripts/reset-gpdb5-cluster.bash [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+rsync -a --delete ./gpdb5-data-copy/ ./gpdb5-data
```
Comment (dgkimura) [inline contrib/pg_upgrade/test/integration/scripts/reset-gpdb5-cluster.bash]: How is this related to current PR?

Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scripts/reset-gpdb5-cluster.bash]: It's not, but it's a useful debugging tool that doesn't really need a full PR and debate.

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c (hunk 2 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Not really.

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c (hunk 3 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: We could do that. I've been trying to let a few real examples drive out the need for extra checks and niceties. 

For example, I've been thinking about how to share some of the assertion logic that is similar between tests while still allowing each test to create their own domain.

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c (hunk 1 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: The rationale for testing in this way is to keep the test description at a high level of abstraction - making it more human readable and hiding the implementation details (queries and such) at a lower level. We can then think at a high level and translate those thoughts into the lower level instructions. 

Hopefully, over more scenarios, we'll find reusable bits for the lower-level abstractions, making them less difficult to write and less error prone.

File: contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test_suite.c [hunk 5]
```diff
@@ -33,9 +38,10 @@ main(int argc, char *argv[])
 	cmockery_parse_arguments(argc, argv);
 
 	const		UnitTest tests[] = {
-		unit_test_setup_teardown(test_an_ao_table_with_data_can_be_upgraded, setup, teardown),
-		unit_test_setup_teardown(test_an_aocs_table_with_data_can_be_upgraded, setup, teardown),
+//		unit_test_setup_teardown(test_an_ao_table_with_data_can_be_upgraded, setup, teardown),
+//		unit_test_setup_teardown(test_an_aocs_table_with_data_can_be_upgraded, setup, teardown),
```
Comment (dgkimura) [inline contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test_suite.c]: Sort of unrelated, but it doesn't seem like cmockery has a good way to run only a single test.  I imagine it's intended more as a unittest framework with fast tests.  Do image that commenting out tests you don't care about is going to be the development workflow or should we start considering other frameworks?

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+static void
+aSubpartitionedHeapTableHasDataInAGpdbFiveCluster(void)
+{
+	PGconn *connection = connectToFive();
+
+	executeQuery(connection, "create schema five_to_six_upgrade;");
+	executeQuery(connection, "set search_path to 'five_to_six_upgrade';");
+	executeQuery(connection, "create table users (id int, age int) distributed by (id) partition by range (id) subpartition by range (age) (partition partition_id start(1) end(3) ( subpartition subpartition_age_first start(1) end(20), subpartition subpartition_age_second start(20) end(30) ))");
+	executeQuery(connection, "insert into users (id, age) values (1, 10), (2, 20)");
+	executeQuery(connection, "vacuum freeze;");
```
Comment (dgkimura) [inline contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Should vacuum freeze be done inside `performUpgrade()` as it seems like a step that must be taken during upgrade that has no direct relation to the individual tests themselves?

File: contrib/pg_upgrade/test/integration/scripts/reset-gpdb5-cluster.bash (hunk 4 above)
Comment (dgkimura) [reply contrib/pg_upgrade/test/integration/scripts/reset-gpdb5-cluster.bash]: Nit: I find adding unrelated changes to PR to be somewhat of a distraction.  If it doesn't need PR then I'd prefer to just push it without PR rather than adding it here.

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c (hunk 1 above)
Comment (dgkimura) [reply contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Okay, fair enough.  I'm good with seeing where this goes, but hope we can keep an eye on complexity as I think that also hurts readability.

File: contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test_suite.c (hunk 5 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/greenplum_five_to_greenplum_six_upgrade_test_suite.c]: I've been thinking about what to do here, but I'm not exactly sure which way to go. I've been considering altering cmockery to allow an argument to focus on a particular test.

Also, I'd like to randomize the suite running order to ensure we don't get test dependencies.

File: contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c (hunk 6 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/subpartitioned_heap_table.c]: Good question. I think we need to investigate exactly why that step is necessary. The previous philosophy of upgrade was to not alter the old cluster during upgrade (beyond hard linking) - which is why I've kept it out of the `performUpgrade()` utility for now.
